examples/exex-subscription/src/main.rs

```diff
@@ -0,0 +1,162 @@
+};
+use reth_ethereum::{
+    exex::{ExExContext, ExExEvent, ExExNotification},
+    node::{api::FullNodeComponents, EthereumNode},
+};
+use std::collections::HashMap;
+use tokio::sync::{mpsc, oneshot};
+use tracing::{error, info};
+
+/// Subscription update format for storage changes.
+/// This is the format that will be sent to the client when a storage change occurs.
+#[derive(Debug, Clone, Copy, Default, serde::Serialize)]
+struct StorageDiff {
+    address: Address,
+    key: U256,
+    old_value: U256,
+    new_value: U256,
+}
+
+/// Subscription request format for storage changes.
+struct SubscriptionRequest {
+    /// The address to subscribe to.
+    address: Address,
+    /// The response channel to send the subscription updates to.
+    response: oneshot::Sender<mpsc::UnboundedReceiver<StorageDiff>>,
+}
+
+/// Subscription request format for storage changes.
+type SubscriptionSender = mpsc::UnboundedSender<SubscriptionRequest>;
+
+/// API to subscribe to storage changes for a specific Ethereum address.
+#[rpc(server, namespace = "watcher")]
+pub trait StorageWatcherApi {
+    /// Subscribes to storage changes for a given Ethereum address and streams `StorageDiff`
+    /// updates.
+    #[subscription(name = "subscribeStorageChanges", item = StorageDiff)]
+    fn subscribe_storage_changes(&self, address: Address) -> SubscriptionResult;
+}
+
+/// API implementation for the storage watcher.
+#[derive(Clone)]
+struct StorageWatcherRpc {
+    /// The subscription sender to send subscription requests to.
+    subscriptions: SubscriptionSender,
+}
+
+impl StorageWatcherRpc {
+    /// Creates a new [`StorageWatcherRpc`] instance with the given subscription sender.
+    fn new(subscriptions: SubscriptionSender) -> Self {
+        Self { subscriptions }
+    }
+}
+
+impl StorageWatcherApiServer for StorageWatcherRpc {
+    fn subscribe_storage_changes(
+        &self,
+        pending: PendingSubscriptionSink,
+        address: Address,
+    ) -> SubscriptionResult {
+        let subscription = self.subscriptions.clone();
+
+        tokio::spawn(async move {
+            let sink = match pending.accept().await {
+                Ok(sink) => sink,
+                Err(e) => {
+                    error!("failed to accept subscription: {e}");
+                    return;
+                }
+            };
+
+            let (resp_tx, resp_rx) = oneshot::channel();
+            subscription.send(SubscriptionRequest { address, response: resp_tx }).unwrap();
+
+            let Ok(mut rx) = resp_rx.await else { return };
+
+            while let Some(diff) = rx.recv().await {
+                let msg = SubscriptionMessage::from_json(&diff).expect("serialize");
+                if sink.send(msg).await.is_err() {
+                    break;
+                }
+            }
+        });
+
+        Ok(())
+    }
+}
+
+async fn my_exex<Node: FullNodeComponents>(
+    mut ctx: ExExContext<Node>,
+    _subscriptions_senderr: SubscriptionSender,
```

Comment:
here we need to actually handle the incoming requests.

we can do this by using a tokio::select over both the notifications and this sender here,
when we receive a new request from the subscription endpoint we need to store this in a new list
through which we then iterate when we receive a new `notification` and only retain (Vec::retain) channels that are still open and send the updates